cpp/src/arrow/flight/sql/server.cc

```diff
@@ -423,6 +494,91 @@ arrow::Result<Result> PackActionResult(ActionCreatePreparedStatementResult resul
   return PackActionResult(pb_result);
 }
 
+arrow::Result<Result> PackActionResult(ActionSetSessionOptionsResult result) {
+  pb::sql::ActionSetSessionOptionsResult pb_result;
+  for (SetSessionOptionResult& res : result.results) {
+    switch (res) {
+      case SetSessionOptionResult::kUnspecified:
+        pb_result.add_results(
+            pb::sql::ActionSetSessionOptionsResult::SET_SESSION_OPTION_RESULT_UNSPECIFIED);
+        break;
+      case SetSessionOptionResult::kOk:
+        pb_result.add_results(
+            pb::sql::ActionSetSessionOptionsResult::SET_SESSION_OPTION_RESULT_OK);
+        break;
+      case SetSessionOptionResult::kInvalidResult:
+        pb_result.add_results(
+            pb::sql::ActionSetSessionOptionsResult::SET_SESSION_OPTION_RESULT_INVALID_VALUE);
+        break;
+      case SetSessionOptionResult::kError:
+        pb_result.add_results(
+            pb::sql::ActionSetSessionOptionsResult::SET_SESSION_OPTION_RESULT_ERROR);
+        break;
+    }
+  }
+  return PackActionResult(pb_result);
+}
+
+arrow::Result<Result> PackActionResult(ActionGetSessionOptionsResult result) {
+  pb::sql::ActionGetSessionOptionsResult pb_result;
+  for (const SessionOption& in_opt : result.session_options) {
+    pb::sql::SessionOption& opt = *pb_result.add_session_options();
```

Comment:
* This does need to be mutable—do you have a preference for alternative code pattern to build nested Proto messages like this?

* I would definitely like shared code, however I felt the best path would be to overhaul all of the Proto<->native object code and factor all of the similar switch-too-much gunk out of the packing/unpacking code and make all of that code more readable with a wholesale shift to factored-out conversion code.  Right now there isn't a common location outside of types.h and overriding Proto constructors from there is kind of ugly factoring as well.

Reply:
The version of the Google C++ guidelines that Arrow was developed with prefers `T*` to `T&`.

Reply:
Also, no Protobuf types can be placed in types.h. They can only be in an internal header.

Reply:
> The version of the Google C++ guidelines that Arrow was developed with prefers `T*` to `T&`.

Can do.  Always believed the consistency of `.` vs `->` was useful, but I will update accordingly, thanks.

Reply:
What I mean is

> Right now there isn't a common location outside of types.h and overriding Proto constructors from there is kind of ugly factoring as well.

We can't refactor anything dealing with Protobuf into a public header. We can have internal headers.

Reply:
> Can do. Always believed the consistency of . vs -> was useful, but I will update accordingly, thanks.

Yeah, it would be nice, but intra-codebase consistency is more valuable for now. (The Google C++ guidelines have since updated to favor the reference in all cases.) But changing this for us would mean hard compatibility breaks in many, many APIs (and it would likely be possible to fix as we've been trying to favor arrow::Result over out parameters _anyways_, but not sure if anyone is driving that refactoring.)